### PR TITLE
[d3d9] Account for vertex declaration size for UP draws

### PR DESCRIPTION
**Pull request by Joshua-Ashton**
The stride may not give us the full picture here as the stride
may not encompass the vertex declaration entirely.

Consider a vertex declaration of size 20, and a stride of 12,
we may not have covered the whole range of space the draw wants with
VertexCount * Stride.

Instead, take (VertexCount - 1) * Stride + VertexDeclSize

Closes: #2046
Closes: #1908

### CODE REVIEW

**Comment by Joshua-Ashton** [comment]
Draft until I hear back about this fixing the issue

File: src/d3d9/d3d9_device.h [hunk 1]
```diff
@@ -1186,6 +1186,13 @@ namespace dxvk {
         : GetHelper(m_state.psConsts);
     }
 
+    inline uint32_t CalculateVertexSizeUP(uint32_t vertexCount, uint32_t stride) {
+      if (unlikely(!vertexCount))
```
Comment (doitsujin) [inline src/d3d9/d3d9_device.h:1190]: can we exit early from `DrawPrimitiveUP` instead if the app is dumb? We shouldn't submit an empty draw to the backend.

**Review by doitsujin — COMMENTED** [review]
Nit, looks fine otherwise. Dumb games be dumb.

**Comment by Joshua-Ashton** [comment]
I wanna take a second look at this tomorrow... Had some bedtime thoughts about it.

**Comment by Joshua-Ashton** [comment]
Ended up looking into this further and what I did here was slightly incorrect, the size of the user data remains the same.

However we need to pad the buffer to this size with 0s instead.

The game has a float3 texcoord trailing just outside the buffer range which causes the whole vertex element to be 0 on NVIDIA, rather than just the .z oob component.

File: src/d3d9/d3d9_device.h (hunk 1 above)
Comment (Joshua-Ashton) [reply src/d3d9/d3d9_device.h]: Done

**Pull request merged**